idx/metric_definition.go

```diff
@@ -0,0 +1,352 @@
+package idx
+
+import (
+	"bytes"
+	"crypto/md5"
+	"fmt"
+	"reflect"
+	"sort"
+	"strings"
+	"unsafe"
+
+	"github.com/raintank/schema"
+	log "github.com/sirupsen/logrus"
+)
+
+//go:generate msgp
+
+type mType uint8
+
+// MType represented by a uint8
+const (
+	MTypeUndefined mType = iota
+	MTypeGauge
+	MTypeRate
+	MTypeCount
+	MTypeCounter
+	MTypeTimestamp
+)
+
+// MetricName stores the name as a []uintptr to strings interned in an object store.
+// Each word is stored as a separate interned string without the '.'. The whole name
+// can be retrieved by calling MetricName.String()
+type MetricName struct {
+	nodes []uintptr
+}
+
+// Nodes returns the []uintptr of interned string addresses
+// for the MetricName
+func (mn *MetricName) Nodes() []uintptr {
+	return mn.nodes
+}
+
+// String returns the full MetricName as a string
+// using data interned in the object store
+func (mn *MetricName) String() string {
+	if len(mn.nodes) == 0 {
+		return ""
+	}
+
+	bld := strings.Builder{}
+	return mn.string(&bld)
+}
+
+func (mn *MetricName) string(bld *strings.Builder) string {
+	// get []int of the lengths of all of the mn.Nodes
+	lns, ok := IdxIntern.Len(mn.nodes)
+	if !ok {
+		internError.Inc()
+		log.Error("idx: Failed to retrieve length of strings from interning library")
+		return ""
+	}
+
+	// should be faster than calling IdxIntern.SetString in a tight loop
+	var tmpSz string
+	szHeader := (*reflect.StringHeader)(unsafe.Pointer(&tmpSz))
+	first, _ := IdxIntern.ObjString(mn.nodes[0])
+	bld.WriteString(first)
+	for idx, nodePtr := range mn.nodes[1:] {
+		szHeader.Data = nodePtr
+		szHeader.Len = lns[idx+1]
+		bld.WriteString(".")
+		bld.WriteString(tmpSz)
+	}
+
+	return bld.String()
+}
+
+// setMetricName interns the MetricName in an
+// object store and stores the addresses of those strings
+// in MetricName.nodes
+func (mn *MetricName) setMetricName(name string) {
+	nodes := strings.Split(name, ".")
+	mn.nodes = make([]uintptr, len(nodes))
+	for i, node := range nodes {
+		// TODO: add error checking? Fail somehow
+		nodePtr, err := IdxIntern.AddOrGet([]byte(node))
+		if err != nil {
+			log.Error(err)
```

Comment:
this should probably `continue` and leave `mn.nodes[i]` at `0`. otherwise that's setting it up for future errors when something tries to convert `mn.nodes[i]` back into a string

Reply:
Upon encountering an error in IdxIntern.AddOrGet the returned value will be 0 and the error, so it is actually setting `mn.nodes[i]` to `0` after this happens.